File: dp/frog.cpp
#include <bits/stdc++.h>

using namespace std;
using vi = vector<int>;


// first jump k0 = 1 unit
// all preceding jumps = k - 1 or k + 1
// stones[i] = distance of stone from start
unordered_map<int, bool> dp;
bool canCross(vi& stones, int pos = 0, int k = 0) {
  int key = pos | k << 11;

  if (dp.count(key) > 0)
      return dp[key];

  for (int i = pos + 1; i < stones.size(); i++) {
      int gap = stones[i] - stones[pos];
      if (gap < k - 1)
          continue;
      if (gap > k + 1)
          return dp[key] = false;
      if (canCross(stones, i, gap))
          return dp[key] = true;
  }

  return dp[key] = (pos == stones.size() - 1);
}

int main() {
  vi stones = { 0,1,3,5,6,8,12,17 };
  canCross(stones);
}

File: dp/frog.py
class Solution:
    dp = {}
    def canCross(self, stones, pos = 0, k = 0):
        """
        :type stones: List[int]
        :rtype: bool
        """
        # the pos is < 1100, so it is small enough for the first 11 bits of an integer
        # the remaining 21 bits are used for k = distance, which also never exceeds 1100
        key = (k << 11) | pos
        
        if (key in self.dp) :
            return self.dp[key]
        
        i = pos 
        for stone in stones[pos + 1:] :
          gap = stone - stones[pos]
          i+=1
          if (gap < k - 1) : 
            continue
          if (gap > k + 1) :
            self.dp[key] = False
            return False
          # else if gap == (k+1 || k-1), recurse
          if (self.canCross(stones, i, gap)) :
            self.dp[key] = True
            return True
        
        self.dp[key] = (pos == (len(stones) - 1))
        return self.dp[key]

s = Solution()
stones = [0,1,2,3,4,8,9,11]
print(s.canCross(stones))
